BUG: using dtype='int64' argument of Series causes ValueError: values cannot be losslessly cast to int64 for integer strings

shubham11941140 opened the pull request:
- [x] closes #44923
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

Added extra check before returning ValueError.

shubham11941140 committed:
Valueerror

shubham11941140 committed:
precommit

jbrockmendel commented on pandas/core/dtypes/cast.py:
```diff
@@ -2096,6 +2096,9 @@ def maybe_cast_to_integer_array(
         )
         return casted
 
+    if all(np.dtype(i) is dtype for i in casted):
```
this is going to be super-slow

jbrockmendel replied:
can maybe update the comment on L2102 if we've found cases that get here

shubham11941140 replied:
I don't recall any cases that reach the ValueError that were there. There was one case which the added code solved.

shubham11941140 replied:
I do not have a faster method to check whether all the cast for all the elements is performed correctly. 

Checking the dtypes is one way but individually we cannot check all elements.

This is executed last, so in many cases we will not reach here. So a bit slow can do.

jreback replied:
yeah we need to be really sure about this check

shubham11941140 replied:
Changed the check as requested.

shubham11941140 committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into b7

shubham11941140 committed:
Optimization

shubham11941140 commented:
As dtype will be an integer dtype there is no need to explicitly check it we can check all integer dtypes.

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -2096,6 +2096,9 @@ def maybe_cast_to_integer_array(
         )
         return casted
 
+    if all(isinstance(i, (int, np.integer)) for i in casted):
```
you can `infer_dtype(casted) == 'integer'` i think

shubham11941140 replied:
Done

shubham11941140 committed:
infer_dtype added

shubham11941140 committed:
precommit

shubham11941140 commented:
@jbrockmendel @jreback any updates?

jbrockmendel commented on pandas/tests/series/test_constructors.py:
```diff
@@ -1810,6 +1810,18 @@ def test_constructor_bool_dtype_missing_values(self):
 
+    def test_constructor_int64_dtype(self):
+        # GH-44923
+        result = Series(["0", "1", "2"], dtype="int64")
```
can you come up with a test case that goes through 2099-2100 but that _shouldn't_ get cast?  maybe ["0", "1", "1.1"]?

shubham11941140 replied:
Series(["0", "1", 0, 1], dtype="int64") this raises the specified ```ValueError```

shubham11941140 replied:
But shouldn't this get cast to ```int64``` as all can be type casted.

shubham11941140 replied:
Should I add the above example and change the code to fit it?

shubham11941140 replied:
@jbrockmendel there is no such case, so it is good to keep that line as it keeps the base covered.

jbrockmendel commented on pandas/core/dtypes/cast.py:
```diff
@@ -2096,6 +2096,9 @@ def maybe_cast_to_integer_array(
         )
         return casted
 
+    if lib.infer_dtype(casted) == "integer":
```
i think this _always_ holds.  The condition we're interested in is whether the casting was lossy.

shubham11941140 replied:
Agreed, but this test currently passes the extra test added.

jreback replied:
do the added tests pass w/o this change?

shubham11941140 replied:
Yes they do.

jbrockmendel replied:
Right, because this condition _always_ holds.

shubham11941140 replied:
But before adding this check the testcase in the issue was not passing.

jbrockmendel replied:
> But before adding this check the testcase in the issue was not passing.

That's a reason to add _some_ check for this, but this particular check is not the right check.  ATM this is equivalent to (but slower than) `if True:`

shubham11941140 commented:
@jreback @jbrockmendel any update on this?

jreback commented:
> @jreback @jbrockmendel any update on this?

@shubham11941140 there is not need to ping. you need to answer @jbrockmendel  last question. I guess you tests is not specific enough (if it already passed)

shubham11941140 commented:
Actually I have added the discussion point to another test case, which might indicate this. I was asking to add another testcase which gives a lossless conversion.

jreback commented:
> Actually I have added the discussion point to another test case, which might indicate this. I was asking to add another testcase which gives a lossless conversion.

great

shubham11941140 commented:
Do I add it?
The last testcase should be correct as the conversion is lossless including this one.

jreback commented:
> Do I add it? The last testcase should be correct as the conversion is lossless including this one.

does that test fail *before* your change? that's the key, does this replicate the original issue.

shubham11941140 commented:
It was currently failing the test.
So I will add it

jreback commented on pandas/tests/series/test_constructors.py:
```diff
@@ -1810,6 +1810,18 @@ def test_constructor_bool_dtype_missing_values(self):
         expected = Series(True, index=[0], dtype="bool")
         tm.assert_series_equal(result, expected)
 
+    def test_constructor_int64_dtype(self):
```
pls parameterize these. also add a case for uint64.

shubham11941140 replied:
Adding a case for uint64 will cause it to crash as this will implicitly type cast to int64 leading to an Assertion Error with uint64.

Parametrization is completed

shubham11941140 committed:
Parametrised

shubham11941140 commented:
@jbrockmendel any update on this? I think I have covered everything.

jreback commented on pandas/tests/series/test_constructors.py:
```diff
@@ -1810,6 +1810,19 @@ def test_constructor_bool_dtype_missing_values(self):
         expected = Series(True, index=[0], dtype="bool")
         tm.assert_series_equal(result, expected)
 
+    @pytest.mark.parametrize("int_dtype", ["int64"])
```
use this fixture instead: any_int_dtype

shubham11941140 replied:
Done

jreback commented on pandas/tests/series/test_constructors.py:
```diff
@@ -1810,6 +1810,19 @@ def test_constructor_bool_dtype_missing_values(self):
+        expected = Series([-1, 0, 1, 2])
+        tm.assert_series_equal(result, expected)
+
+    def test_constructor_float64_dtype(self):
```
use any_float_dtype

shubham11941140 replied:
Done

shubham11941140 committed:
changed to any_dtype

jreback commented on pandas/tests/series/test_constructors.py:
```diff
@@ -1810,6 +1810,20 @@ def test_constructor_bool_dtype_missing_values(self):
         tm.assert_series_equal(result, expected)
 
+    @pytest.mark.parametrize("any_int_dtype", ["int64"])
+    def test_constructor_int64_dtype(self, any_int_dtype):
```
no, pls just use the fixture itself, e.g. no parameterize

shubham11941140 replied:
This is causing Assertion Error.

shubham11941140 replied:
![image](https://user-images.githubusercontent.com/63910248/147833002-384bb60d-c65d-469c-b43e-e70361969744.png)

shubham11941140 replied:
The previous code segment is leading to this issue, if we have only int64 there is no issue.

jreback replied:
you need to match the expected value as well

shubham11941140 replied:
I have added them in the parameters which are passing.

uint -> uint8, uint16, uint32, uint64 are failing due to internal code implementation
Float -> Float32, Float64 are failing as it is unable to implicitly cast strings.

The other have been added into the parametrize function defined, which are not causing any problem.

jreback replied:
u can xfail the failing ones that are tricky to fix eg Float but the others should work

shubham11941140 replied:
![image](https://user-images.githubusercontent.com/63910248/147844312-6ecc6742-086a-4e03-8b56-07c239fc7dae.png)

shubham11941140 replied:
I am getting this for uint.

shubham11941140 replied:
@jreback is this fine?

shubham11941140 replied:
@jreback I think I have covered everything?

jreback replied:
@shubham11941140 you are not using the fixtures pls do so

jreback replied:
just remove the paramterize completely

shubham11941140 replied:
uint -> uint8, uint16, uint32, uint64 are failing due to internal code implementation. Do i fix this?

shubham11941140 replied:
@jreback removed parametrization, now it should be ready.

jreback commented on pandas/tests/series/test_constructors.py:
```diff
@@ -1810,6 +1810,20 @@ def test_constructor_bool_dtype_missing_values(self):
+        expected = Series([-1, 0, 1, 2])
+        tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize("any_float_dtype", ["float64"])
```
same

shubham11941140 replied:
It is the same as the previous image.

shubham11941140 committed:
Added xfail

shubham11941140 committed:
precommit

shubham11941140 committed:
Removed parametrise

shubham11941140 committed:
precommit fix

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -2062,8 +2062,17 @@ def maybe_cast_to_integer_array(
-        raise OverflowError("Trying to coerce negative values to unsigned integers")
+    if is_unsigned_integer_dtype(dtype):
+        try:
+            if (arr < 0).any():
```
why is this is a try except

shubham11941140 replied:
The input in arr is ["0", "1", "2"], this is causing a TypeError in the check ```(arr < 0).any()``` but it can be casted to uint, so I am checking ```(casted < 0).any()``` which it hits and casts correctly

jreback replied:
this doesn't make sense
why does this check matter? is the casted valid? 

where is the test that checks the overflow?

strive for minimal code

shubham11941140 replied:
```result = Series(["0", "1", "2"], dtype=uint8)```, this should give a valid cast and casted is valid, but ```(arr < 0).any()``` is giving TypeError.

shubham11941140 replied:
The point is that string arrays cannot be compared with integers, numpy does not support it, thus it is leading to a Type Error.

jreback commented on pandas/tests/series/test_constructors.py:
```diff
@@ -1810,6 +1810,20 @@ def test_constructor_bool_dtype_missing_values(self):
+
+    def test_constructor_float64_dtype(self, any_float_dtype):
+        # GH-44923
+        if any_float_dtype in ["Float32", "Float64"]:
```
is there an issue for this?

shubham11941140 replied:
Float -> Float32, Float64 are failing as it is unable to implicitly cast strings. As the implicit cast fails so I am xfailing them, 

#### ```u can xfail the failing ones that are tricky to fix eg Float but the others should work```

This one you had mentioned above.

jbrockmendel replied:
i think these now work in main #45424

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -2062,8 +2062,17 @@ def maybe_cast_to_integer_array(
+                )
+        except TypeError:
+            if (casted < 0).any():
+                raise OverflowError(
```
where is this hit in tests?

shubham11941140 replied:
```result = Series(["0", "1", "2"], dtype=uint8)```, this is hit. For all uint in ```any_int_dtype```

jbrockmendel replied:
i dont understand this.  if i pass `["0", "1", "2"]` to np.array on L2046 with dtype=np.uint8 i get back a uint8 ndarray that shouldn't raise here

jreback reviewed (changes requested):
you are adding additional checks here? show what case needs these

shubham11941140 commented:
@jreback any update?

jreback commented:
> @jreback any update?

@shubham11941140 you are touching some very particular code, i don't know how to proceed here. this PR keeps expanding in scope & code changes w/o test cases.

shubham11941140 commented:
I will explain it here in detail. It is the same test case in the ```test_constructors.py``` file

```result = Series(["0", "1", "2"], dtype=any_int_dtype)```
```exp = Series([0, 1, 2], dtype=any_int_dtype)```

```any_int_dtype``` covers ```uint8``` and every other unsigned integer dtype

So, taking ```uint8```:

```result = Series(["0", "1", "2"], dtype=uint8)```
```exp = Series([0, 1, 2], dtype=uint8)```

I have written these 2 cases and they must be equal i.e. ```tm.assert_frame_equal``` must hold.

The issue is that when dtype=uint8 or for the matter of fact any dtype where ```is_unsigned_integer_dtype(dtype)``` holds, we reach the condition of ```(arr < 0).any()```.

If you look very carefully, ```arr = ["0", "1", "2"]```, our ```np.ndarray```, that contains strings. The function ```(arr < 0).any()``` is a numpy function that DOES NOT ALLOW the comparison of strings to integers (As "0" is a ```str``` and 0 is ```int```) and leads to a ```TypeError```.

However, there is no error in this case as strings can be implicitly cast to int by this function.

This is done within the casted variable which is ```casted = [0, 1, 2]``` which is then casted into ```uint``` dtype. If we apply the function ```(casted < 0).any()```, there should not be any error for the above specified testcase.

For this testcase as specified, I have created a try except block that prevents other tests from breaking and allows type casting of strings to unsigned integers without running into a ```TypeError``` of the NumPy function.

In this manner, I am not running into an issue with the ```unsigned_integer_dtype``` and can perform the implicit cast.

I hope this can now explain the code change I have done associated with the test case. If you are unable to understand anything further. I can schedule a call to explain it. 

@jreback

shubham11941140 commented:
@jreback do I need to give a more elaborate explanation?

shubham11941140 commented:
@jbrockmendel @jreback any update?

jbrockmendel commented:
> any update?

The check on L2108 still needs to be changed to something meaningful.  The check is about the casting not being lossy.

shubham11941140 commented:
How do I check whether it is a lossless change? Is there any construct?

jreback commented:
> How do I check whether it is a lossless change? Is there any construct?

you need to construct a test explicitly for this

shubham11941140 committed:
removed obvious line

shubham11941140 commented:
@jreback @jbrockmendel . I have removed the obvious statement used for the return condition and added to the above condition.

This is obviously checking the lossless condition and it is combined with the Overflow condition as previously written as if the user gives very large strings to be casted into small integers such as ```int8```, we will not run into a similar issue again.

I think this will satisfy the requisite conditions now.

jbrockmendel commented on pandas/core/dtypes/cast.py:
```diff
@@ -2074,7 +2083,7 @@ def maybe_cast_to_integer_array(
         raise ValueError("Trying to coerce float values to integers")
 
-    if casted.dtype < arr.dtype:
+    if casted.dtype < arr.dtype or lib.infer_dtype(casted) < lib.infer_dtype(arr):
```
`lib.infer_dtype(casted) < lib.infer_dtype(arr)` im not sure what this is supposed to mean

jbrockmendel commented on pandas/core/dtypes/cast.py:
```diff
@@ -2062,8 +2062,17 @@ def maybe_cast_to_integer_array(
+                raise OverflowError(
+                    "Trying to coerce negative values to unsigned integers"
+                )
+        except TypeError:
```
can you add a comment here about what cases get here

shubham11941140 commented:
```lib.infer_dtype(casted) < lib.infer_dtype(arr)``` it is the same like the extension of ```casted.dtype < arr.dtype``` to inferred dtypes as they are not properly initialized in the input. However they check the same as the condition of the lossless casting as the warning message means.

jbrockmendel commented:
> lib.infer_dtype(casted) < lib.infer_dtype(arr) it is the same like the extension of casted.dtype < arr.dtype to inferred dtypes as they are not properly initialized in the input. However they check the same as the condition of the lossless casting as the warning message means.

lib.infer_dtype returns a string.  Why is comparing two strings useful here?

shubham11941140 committed:
Checking string dtype directly

shubham11941140 commented:
if the dtype is a string it is a situation of lossless cast, so that can be specified directly.

jreback commented:
@shubham11941140 if you want to merge master and update to comments

shubham11941140 committed:
Merge branch 'main' into b7

shubham11941140 commented:
Branch is updated.

jreback reviewed (changes requested):
@shubham11941140 before you do any patches here, can you demonstrate failing cases in tests

shubham11941140 commented:
I did not understand what do you want me to do?

shubham11941140 commented:
On running the query, it mentions a FutureWarning when comparing ```return bool(asarray(a1 == a2).all())```.

This kind of error is not coming from the changes I have made.

shubham11941140 commented:
The issue is that string comparison with integer is going past the limit of the NumPy comparison because of which it is failing.

shubham11941140 commented:
![image](https://user-images.githubusercontent.com/63910248/157061096-ff83adba-7163-40da-8d31-f7d0f480a0d7.png)

The strange part is that this should be raising a Warning with the ```np.array_equal``` comparison.

Very strange is that numpy dev is raising an error which is not happening on my machine.

shubham11941140 commented:
I have tested the build on my local machine and it is not failing, but 2 errors come here.

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in main, address the comments in the code diff and we can reopen
